### DISCUSSION TITLE
error Error [TypeError]: request.formData is not a function

### DISCUSSION THREAD

**Discussion by loganknecht** [vercel/next.js#53827] [Help] [answered]
### Summary

# Problem

I am uploading a file using the `FormData` object.

I cannot call `request.formData()` because it says it is not a function.

# Package JSON Dependencies
```json
{
    // ...
    "dependencies": {
        "@emotion/react": "^11.11.1",
        "@emotion/styled": "^11.11.0",
        "@mui/material": "^5.13.4",
        "@types/file-saver": "^2.0.5",
        "@types/node": "20.2.5",
        "@types/react": "18.2.9",
        "@types/react-dom": "18.2.4",
        "@types/validator": "^13.7.17",
        "eslint": "8.42.0",
        "eslint-config-next": "13.4.4",
        "file-saver": "^2.0.5",
        "next": "^13.4.13",
        "react": "18.2.0",
        "react-dom": "18.2.0",
        "replace-in-file": "^7.0.1",
        "typescript": "5.1.3",
        "validator": "^13.9.0"
    }
    // ...
}
```

# Client Code

```TypeScript
const form_data = new FormData();

// Required Data
form_data.append("application_id", current_application_id_string);
form_data.append("application_name", current_application_name_string);
form_data.append("application_url", current_application_url_string);

// Optional Data
if (form_state.application_icon !== null) {
    form_data.append(
        "application_icon",
        form_state.application_icon,
        form_state.application_icon.name
    );
}

const api_url = `${api_hostname}/${api_path}`;

const fetch_promise = fetch(api_url, {
    method: "POST",
    headers: {
        "Content-Type": "multipart/form-data",
    },
    body: form_data,
});

fetch_promise
    .then((response) => {
        // ...
    })
    .catch((error: any) => {
        // ...
    });
```

# Server Code

```TypeScript
import { NextApiRequest, NextApiResponse } from "next";

async function submit_package_request(request: NextApiRequest, response: NextApiResponse) {
    console.log(request_data);
    const request_data = await request.formData();
    // ...
}
```

### Additional information

```js
I feel like I'm taking crazy pills.

I see a lot of references to "just use formidable". I'm open to that avenue, but I also see people saying "just use request.formData"

I cannot find documentation about the `request.formData()` approach.
And I can't find anything to indicate what the issue is here.

I have updated to the latest `Next` library. My code looks sane.

What am I doing wrong?

Why can I not get this file on the API side?
```

**Comment by loganknecht** [comment, marked as answer]
Ok... So this is a huge headache.

For anyone who finds this. The issue was several things.

1. Next 13 has a different support system for routes which uses route handlers
    - https://nextjs.org/docs/app/building-your-application/routing/route-handlers
2.  You cannot use the `request.formData` without using `route handlers`
3. You need to also make sure your route handler is exported correctly
4. Even with these pieces in place, you still need to disable the body parser

In order to fix this I had to move my api endpoints from the `repo_root_dir/src/pages/api/v1/package/submit.tsx` directory to `repo_root_dir/src/app/api/v1/package/submit/route.ts`

THEN

I had to rename my function to `POST`

AND THEN I had to set the configuration to not use the body parser

```typescript
import { NextApiRequest, NextApiResponse } from "next";

export const config = {
    api: {
        bodyParser: false,
    },
};

export async function POST(request: NextApiRequest, response: NextApiResponse) {
    console.log(request_data);
    const request_data = await request.formData();
    // ...
}
```

If I can say anything about this whole experience is that the NextJS team is doing a terrible job with documentation and their migration strategy for these changes. It has been an incredibly frustrating  experience to chase this down and even more so to debug and triage this was a nightmare.

Good luck to anyone else who finds themselves in this situation as well.

**Reply by AidanGuarniere** [reply]
lifesaver. thank you

**Reply by boris-szl** [reply]
my god, thank you so much, I have been fighting with this for a week..

**Reply by everurstruly** [reply]
Still breaks from time to time on my end. I'm frustrated to say the least. Been having back and forth with Appwrite support just to upload a dang file/blob object

**Comment by tamvo22** [comment]
I ran into this issue as well. However, I was able to get it working with the help of the formidable-serverless library.

```
import formidable from 'formidable-serverless';
import { NextApiRequest, NextApiResponse } from "next";

export const config = {
  api: {
    bodyParser: false,
  },
};

async function submit_package_request(request: NextApiRequest, response: NextApiResponse) {
      const form = new formidable.IncomingForm();
      form.parse(request, async (err, fields, files) => {
         console.log(err, fields, files);
      });
}

```

**Reply by loganknecht** [reply]
Good information, but misses the intention.

Next 13 supports this by default now if you follow the approach I advise.

**Reply by tamvo22** [reply]
That's correct Next 13 supports formData() by default. I forgot to mention that this issue is related Next12 or lower.

**Comment by VikashLoomba** [comment]
I don't mean to be ignorant of the complexity here, but why is not possible to handle FormData natively within the standard `/pages/api` directory?

**Reply by zhangxinyong12** [reply]
How to solve it. I also had a problem with consent

**Reply by loganknecht** [reply]
NextJS13+ is different and allows for its usage. Prior to that this wasn't supported. It's because the library implementation on the api side was enhanced when they did the migration to the new directory structure and features.

**Comment by joras-droid** [comment]
literally, i stucked in this from 2 days. godddd 
By the way, thank you so much   @loganknecht loganknecht
finally solved the problem

**Comment by TiagoJeronimo** [comment]
I'm using`request.formData()`, but unfortunately, I keep getting the following warning: `ExperimentalWarning: buffer.File is an experimental feature and might change at any time`. 
Any idea how I can suppress this warning? Or should I use some other library like `formidable-serverless` for now?

**Reply by Nazehs** [reply]
I am also getting the same error with `formidable-serverless`  and I am wondering why upload is now a big deal with nextjs. I am having Having the below code

```import type { NextApiRequest, NextApiResponse } from 'next';
import formidable from 'formidable-serverless';
import { v2 as cloudinary } from 'cloudinary';

cloudinary.config({
    cloud_name: process.env.CLOUDINARY_NAME,
    api_key: process.env.CLOUDINARY_API_KEY,
    api_secret: process.env.CLOUDINARY_API_SECRET,
    secure: true,
});

export const config = {
    api: {
        bodyParser: false,
    },
};

type Data = {
    message: string;
    url?: string;
};

const handler = async (req: NextApiRequest, res: NextApiResponse<Data>) => {
    if (req.method !== 'POST') {
        res.setHeader('Allow', ['POST']);
        res.status(405).end('Method ' + req.method + ' Not Allowed');
        return;
    }

    const form = new formidable.IncomingForm();

    form.parse(req, async (err: any, fields: any, files: any) => {
        if (err) {
            console.error(err);
            res.status(500).json({ message: 'Error parsing the form data.' });
            return;
        }

        const uploadType: string = fields.uploadType;
        const userId = fields.userId;

        if (!uploadType) {
            res.status(400).json({ message: 'Upload type is required.' });
            return;
        }

        let folderPath = '';

        switch (uploadType) {
            case 'avatar':
                folderPath = 'avatars';
                break;
            case 'postContent':
                folderPath = 'posts';
                break;
            default:
                res.status(400).json({ message: 'Invalid upload type.' });
                return;
        }

        if (!files.file) {
            res.status(400).json({ message: 'File upload error: No file was uploaded.' });
            return;
        }

        const file = files.file;

        try {
            const result = await cloudinary.uploader.upload(file.filepath, {
                folder: folderPath,
                public_id: uploadType === 'avatar' ? `${userId}` : undefined,
                overwrite: uploadType === 'avatar',
            });

            res.status(200).json({ message: 'File uploaded successfully.', url: result.secure_url });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Error uploading the file.' });
        }
    });
};

export default handler;
```

**Comment by einazare** [comment]
I am having the same issue. I am using NextJS `14.0.4` with Pages Router, and the above solutions do not work for me.
Any idea how I could access the form data that was sent from the client side? 
I am trying to read a media file, that I want to process, but I do not know how I can read the file on the server side :(

**Comment by jamesleeht** [comment]
There's another issue related to this. The default fetch API actually accepts a `Request()` object:
https://developer.mozilla.org/en-US/docs/Web/API/Request

However, if you set `formData` in a `Request()` and pass it to fetch, fetch just drops the form data. Took me a long while to find this related issue.

**Comment by ckgardner** [comment]
I'm on Next 14, and this is still an issue. `req.formData()` is not a function. The body parser trick isn't doing anything.

**Reply by matheuseabra** [reply]
Same here

**Comment by collinsmilimo** [comment]
Incase there is any one still stuck with this... 
I came to realize the request goes through when using postman and fails when using axios, you should switch to fetch and maintain your backend route as below


import { NextApiRequest, NextApiResponse } from "next";

export async function POST(request: NextApiRequest, response: NextApiResponse) {
    console.log(request_data);
    const request_data = await request.formData();
    // ...
}

**Comment by m-shum** [comment]
Next 15 with app router and... you guessed it, still running into this problem using a server action and a route handler. 
For reference, the flow is something like this:

`Client form component:`
```
'use client'

import { useActionState } from 'react'
import { login } from '@/actions'

export default function LoginForm() {
  const [state, formAction, pending] = useActionState(login, initialState)
  return <form action={formAction}>....</form>
}

   ```
   
`Then a server action to handle form submission`
```
'use server'

import { redirect } from 'next/navigation'
import { cookies } from 'next/headers'

import type { TFormState } from '@/types'

export async function login(
  prevState: TFormState,
  formData: FormData
): Promise<TFormState> {
  const res = await fetch(`${process.env.URL}/api/login`, {
    method: 'POST',
    body: formData,
  })

  const data = await res.json()
  if (res.ok) {
   // handle cookie stuff etc
    redirect(`/users/${data.id}`)
  } else
    return {
      message: data.error,
      status: res.status,
    }
}
```

`And finally an API route:`
```
import { NextResponse } from 'next/server'
import { users } from '@/users'

export async function POST(req: Request): Promise<void | Response> {
  const formData = await req.formData() // <-- this right here throws an error in Vercel: "Failed to parse body as FormData"
  const email = formData.get('email')
  const password = formData.get('password')
  const rememberUser = formData.get('rememberUser')

 // fetch user, throw errors, write a love poem to skynet, whatever
 
  return new NextResponse() // whatever response
}
```

There's a chance I'm doing something very stupid, in which case I'm all ears because this is driving me up the wall.
